### ISSUE TITLE
Placing Guardian tokens inside HttpOnly cookies and reading them from there

### ISSUE THREAD

**Issue by egeersoz** [ueberauth/guardian#278]
[I posted this on Stackoverflow](http://stackoverflow.com/questions/42796602/placing-guardian-tokens-inside-cookies-and-reading-them-from-there), but I figured I would ask it here as well.

Essentially, I'm trying to store Guardian jwts inside `HttpOnly` and `Secure` cookies when sending them to the front-end, and performing the verification by reading them from those cookies for incoming requests (instead of from the auth header).

Does Guardian currently support this? I read the [simple guardian](http://blog.overstuffedgorilla.com/simple-guardian/) post but couldn't find anything specific. Apologies if I missed it.

My issue with reading the jwt from Authorization header is that this (afaik) requires the client to be able to read it (with JavaScript), which makes the approach inherently unsafe if the client application is a browser (which should always be considered a hostile environment).

**Comment by hassox** [comment]
Hey. I believe what you're describing is session auth. In your browser pipeline if you use the verify session plug it should pick it up. You'll also need to use the Guardian.sign_in function to make sure your token goes into the session.

Please let me know if that is unclear.

**Comment by hassox** [comment]
Are you using an spa using fetch, maybe even over CORS? You may need to tell the call to fetch to include the credentials

**Comment by egeersoz** [comment]
Yes, using an SPA, with [CORSPlug](https://github.com/mschae/cors_plug) configured on the endpoint to accept the SPA's origin.

I included the credentials in the http call, but still getting 401 - Unauthenticated. Maybe there are no credentials to include? I mean the browser doesn't have any cookies, so I'm not sure where it would get the credentials from.

I don't think we're going in the right track. [Here's the conn at the time of the server's response](https://gist.github.com/egeersoz/8fc08f691ce709aa48a8242bc87487e9). The jwt gets inserted into the session, but note that the cookies map is empty.

**Comment by hassox** [comment]
Can you show the chrome/safari request/response headers you're seeing on your client? (from the dev environment) in the network headers tab? I'd like to see if it's getting to the client. From memory sessions are flushed just before the response is sent in plug.

**Comment by egeersoz** [comment]
Login response from the server:

```
HTTP/1.1 200 OK
server: Cowboy
date: Wed, 15 Mar 2017 01:20:37 GMT
content-length: 59
set-cookie: _myapp_key=SFMyNTY.g3QAAAABbQAAABBndWFyZGlhbl9kZWZhdWx0bQAAAUNleUpoYkdjaU9pSklVelV4TWlJc0luUjVjQ0k2SWtwWFZDSjkuZXlKaGRXUWlPaUpWYzJWeU9qSWlMQ0psZUhBaU9qRTBPVEl4TXpJNE16Y3NJbWxoZENJNk1UUTRPVFUwTURnek55d2lhWE56SWpvaVFYTndhWEpsSWl3aWFuUnBJam9pTlRBMU1tRXdZamd0TmpBeE1DMDBOV015TFdKa1pEVXRaVE15TnpZMFlqSTFaRGN3SWl3aWNHVnRJanA3ZlN3aWMzVmlJam9pVlhObGNqb3lJaXdpZEhsd0lqb2lZV05qWlhOekluMC4xSkRrQzBhN2RkT2F6TTNKRnJ3bGVQdkFPczM5U0paQWZwdHFaSVlNSGNBUm5peDBybWZPaDBjdlhYOEtpRjR5all0M0VHUExFTnZDZHBuMFpqSEVPQQ.aGLE3Lf_7NJfmtE7y5tH31CMZTlckL37UEf9iOFmIag; path=/; HttpOnly
content-type: application/json; charset=utf-8
cache-control: max-age=0, private, must-revalidate
x-request-id: f4r7sard8i472555gsfgckf2p4fg3uvn
access-control-allow-origin: http://localhost:8080
access-control-expose-headers: 
access-control-allow-credentials: true
vary: Origin
```

Following request to /api/account:

```
GET /api/account/ HTTP/1.1
Host: localhost:4000
Connection: keep-alive
Accept: application/json
Origin: http://localhost:8080
User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_3) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/56.0.2924.87 Safari/537.36
Referer: http://localhost:8080/
Accept-Encoding: gzip, deflate, sdch, br
Accept-Language: en-US,en;q=0.8,tr;q=0.6
```

Headers of the 401 Response from the server:

```
HTTP/1.1 401 Unauthorized
server: Cowboy
date: Wed, 15 Mar 2017 01:27:31 GMT
content-length: 30
cache-control: max-age=0, private, must-revalidate
x-request-id: 83js05m5qillugq961dd1qt35efpvofc
access-control-allow-origin: http://localhost:8080
access-control-expose-headers: 
access-control-allow-credentials: true
vary: Origin
content-type: application/json; charset=utf-8
```

**Comment by hassox** [comment]
Thanks so much. So looking at the cookie that you've provided that actually made it back from the sign in response I pulled it apart to see what was there.

```ruby
irb(main):001:0> str = "g3QAAAABbQAAABBndWFyZGlhbl9kZWZhdWx0bQAAAUNleUpoYkdjaU9pSklVelV4TWlJc0luUjVjQ0k2SWtwWFZDSjkuZXlKaGRXUWlPaUpWYzJWeU9qSWlMQ0psZUhBaU9qRTBPVEl4TXpJNE16Y3NJbWxoZENJNk1UUTRPVFUwTURnek55d2lhWE56SWpvaVFYTndhWEpsSWl3aWFuUnBJam9pTlRBMU1tRXdZamd0TmpBeE1DMDBOV015TFdKa1pEVXRaVE15TnpZMFlqSTFaRGN3SWl3aWNHVnRJanA3ZlN3aWMzVmlJam9pVlhObGNqb3lJaXdpZEhsd0lqb2lZV05qWlhOekluMC4xSkRrQzBhN2RkT2F6TTNKRnJ3bGVQdkFPczM5U0paQWZwdHFaSVlNSGNBUm5peDBybWZPaDBjdlhYOEtpRjR5all0M0VHUExFTnZDZHBuMFpqSEVPQQ"
=> "g3QAAAABbQAAABBndWFyZGlhbl9kZWZhdWx0bQAAAUNleUpoYkdjaU9pSklVelV4TWlJc0luUjVjQ0k2SWtwWFZDSjkuZXlKaGRXUWlPaUpWYzJWeU9qSWlMQ0psZUhBaU9qRTBPVEl4TXpJNE16Y3NJbWxoZENJNk1UUTRPVFUwTURnek55d2lhWE56SWpvaVFYTndhWEpsSWl3aWFuUnBJam9pTlRBMU1tRXdZamd0TmpBeE1DMDBOV015TFdKa1pEVXRaVE15TnpZMFlqSTFaRGN3SWl3aWNHVnRJanA3ZlN3aWMzVmlJam9pVlhObGNqb3lJaXdpZEhsd0lqb2lZV05qWlhOekluMC4xSkRrQzBhN2RkT2F6TTNKRnJ3bGVQdkFPczM5U0paQWZwdHFaSVlNSGNBUm5peDBybWZPaDBjdlhYOEtpRjR5all0M0VHUExFTnZDZHBuMFpqSEVPQQ"

irb(main):009:0> Base64.decode64(str)
=> "\x83t\x00\x00\x00\x01m\x00\x00\x00\x10guardian_defaultm\x00\x00\x01CeyJhbGciOiJIUzUxMiIsInR5cCI6IkpXVCJ9.eyJhdWQiOiJVc2VyOjIiLCJleHAiOjE0OTIxMzI4MzcsImlhdCI6MTQ4OTU0MDgzNywiaXNzIjoiQXNwaXJlIiwianRpIjoiNTA1MmEwYjgtNjAxMC00NWMyLWJkZDUtZTMyNzY0YjI1ZDcwIiwicGVtIjp7fSwic3ViIjoiVXNlcjoyIiwidHlwIjoiYWNjZXNzIn0.1JDkC0a7ddOazM3JFrwlePvAOs39SJZAfptqZIYMHcARnix0rmfOh0cvXX8KiF4yjYt3EGPLENvCdpn0ZjHEOA"

irb(main):010:0> payload = "eyJhdWQiOiJVc2VyOjIiLCJleHAiOjE0OTIxMzI4MzcsImlhdCI6MTQ4OTU0MDgzNywiaXNzIjoiQXNwaXJlIiwianRpIjoiNTA1MmEwYjgtNjAxMC00NWMyLWJkZDUtZTMyNzY0YjI1ZDcwIiwicGVtIjp7fSwic3ViIjoiVXNlcjoyIiwidHlwIjoiYWNjZXNzIn0"

irb(main):011:0> Base64.decode64(payload)
=> "{\"aud\":\"User:2\",\"exp\":1492132837,\"iat\":1489540837,\"iss\":\"Aspire\",\"jti\":\"5052a0b8-6010-45c2-bdd5-e32764b25d70\",\"pem\":{},\"sub\":\"User:2\",\"typ\":\"access\"}"
```

I don't believe you have a problem getting the session from the server to the client. It seems to be the other way.

When using `fetch` and you want to use the session with cors I've always found that I need to explicitly instruct `fetch` to include the credentials.

https://fetch.spec.whatwg.org/#cors-protocol-and-credentials
```javascript
fetch("./", { credentials:"include" }).then(/* … */)
```

Can you please try this in your client and let me know how you go?

**Comment by egeersoz** [comment]
Yes, I'm already doing that (not with Fetch, but rather [Vue Resource](https://github.com/pagekit/vue-resource/blob/develop/docs/http.md)):

`this.$http.get("http://localhost:4000/api/account/", {credentials: true}).then(/* ... */)`

That's getting me 401 unauthorized.

**Comment by egeersoz** [comment]
If I remove `Plug.Guardian.EnsureAuthenticated`, I don't get any errors, but the resulting response has empty data, and the user I'm trying to extract from the conn using `Guardian.Plug.current_resource(conn)` is empty.

```
def show(conn, _) do
    user = Guardian.Plug.current_resource(conn)
    Logger.debug inspect user
    render(conn, "show.json", user: user)
  end
```

**Comment by hassox** [comment]
ok agreed that this is strange :\ 

Can you add a debugger plug into your pipeline just after the fetch session and dump out the parsed session? It'd be helpful to know if it's getting back on the way in before we get to the guardian plugs.

Sorry that you're having trouble :(

**Comment by egeersoz** [comment]
How do I add a debugger plug into the pipeline? :)

**Comment by hassox** [comment]
Oh sorry... Just a simple one to splat things out. Something like:

```elixir

defmodule DebugMe do
  def init(_), do: :ok

  def call(conn) do
     Logger.debug inspect conn.req_headers
     Logger.debug inspect conn.req_cookies
     Logger.debug inspect Map.get(conn.private, :plug_session)

     conn
  end
end
```

Then throw that in your pipeline

```elixir
pipeline :api do
    plug :accepts, ["json"]
    plug :fetch_session
    plug DebugMe
    plug Guardian.Plug.VerifySession
    plug Guardian.Plug.EnsureAuthenticated
    plug Guardian.Plug.LoadResource
  end
```

**Comment by egeersoz** [comment]
Alright, debug data immediately after `:fetch_session` - [here you go](https://gist.github.com/egeersoz/0c272860a839e0d444b4527a2e65250c).

**Comment by hassox** [comment]
Thanks. That looks very strange. At the point you've dumped guardian sign in has already run. You can see it in the session with the flag to write.

After that request, what is the debug information? i.e. the next request that you're expecting to be logged in.

**Comment by egeersoz** [comment]
Whoops, sorry. I think I provided the wrong debug log. I restarted the server and re-ran the steps. [This is what I got](https://gist.github.com/egeersoz/1d4fb9d9cada452d003b9bffb4ce8553).

**Comment by hassox** [comment]
Thanks for the update.

I see there that there is no cookie header or request cookies. I went back and had a look at the request log again. There is no cookie being sent to the server in the request :(

If you looks at a request from this page for ex, you can see that there is a cookie included in the request headers. I'm not seeing any evidence that the client is sending cookies back up to the server :\

**Comment by egeersoz** [comment]
Yeah, hmm. Might it be because the cookie has the `HttpOnly` option set?

**Comment by hassox** [comment]
Maybe... I'm not sure how to deal with that :\

**Comment by egeersoz** [comment]
No worries. I think I'm going to go back to the regular approach. Thanks a lot for your help! :)

**Comment by hassox** [comment]
Np. Good luck!

**Comment by jadeddev** [comment]
Hey, I am having a similar problem to @egeersoz 's original problem. I was wondering if anyone resolved being able to pass httponly tokens through an api pipeline?

**Comment by hassox** [comment]
If you're using a cookie that is not the storage you should write your own Verify* style middleware. Those middlewares just find a token from some location, validate them and add them to the Conn struct. It should be fairly straightforward to implement one for your usecase

**Comment by hassox** [comment]
Provided the client actually sends it ;)

**Comment by jadeddev** [comment]
Hey, apologies but I'm not 100% what you mean in your first message.

**Comment by hassox** [comment]
Yeah good call it didn't come out too well. So if you're using cookies, they're not one of the out of the box supported ways to get a token into the system. In this case you need your own middleware to locate the token and call verify on it. If it's valid put it in the Conn struct like the Verify* modules do. You can refer to the header or session Verify modules for an example of an implementation.

**Comment by jadeddev** [comment]
@hassox thanks, will dig a little further :)

**Comment by hassox** [comment]
@connorjacobsen not much to go on mate. Are you using cors?
